File: lab5/server/src/streebog.c
#include "streebog.h"

#include <stddef.h>
#include <stdint.h>
#include <string.h>

#include "streebog_constants.h"

#define STREEBOG_ROUNDS 12

static void streebog_xor(const uint8_t* a, const uint8_t* b, uint8_t* out) {
    for (size_t i = 0; i < STREEBOG_BLOCK_SIZE; ++i) {
        out[i] = (uint8_t)(a[i] ^ b[i]);
    }
}

static void streebog_add(const uint8_t* a, const uint8_t* b, uint8_t* out) {
    uint16_t carry = 0;
    for (size_t i = 0; i < STREEBOG_BLOCK_SIZE; ++i) {
        uint16_t sum = (uint16_t)a[i] + (uint16_t)b[i] + carry;
        out[i] = (uint8_t)(sum & 0xFFU);
        carry = sum >> 8;
    }
}

static void streebog_s(const uint8_t* in, uint8_t* out) {
    for (size_t i = 0; i < STREEBOG_BLOCK_SIZE; ++i) {
        out[i] = STREEBOG_PI[in[STREEBOG_BLOCK_SIZE - 1U - i]];
    }
}

static void streebog_p(const uint8_t* in, uint8_t* out) {
    for (size_t i = 0; i < STREEBOG_BLOCK_SIZE; ++i) {
        out[i] = in[STREEBOG_TAU[STREEBOG_BLOCK_SIZE - 1U - i]];
    }
}

static void streebog_l(const uint8_t* in, uint8_t* out) {
    memset(out, 0, STREEBOG_BLOCK_SIZE);
    for (int i = 7; i >= 0; --i) {
        for (int n = 0; n < 8; ++n) {
            int p = 63;
            uint8_t* target = &out[i * 8 + n];
            for (int j = 7; j >= 0; --j) {
                uint8_t byte = in[i * 8 + j];
                for (int k = 0; k < 8; ++k) {
                    if ((byte >> k) & 0x01U) {
                        *target ^= STREEBOG_L[(size_t)p * 8 + (size_t)n];
                    }
                    --p;
                }
            }
        }
    }
}

static void streebog_spl(uint8_t* state) {
    uint8_t tmp1[STREEBOG_BLOCK_SIZE];
    uint8_t tmp2[STREEBOG_BLOCK_SIZE];
    streebog_s(state, tmp1);
    streebog_p(tmp1, tmp2);
    streebog_l(tmp2, state);
}

static void streebog_e(const uint8_t* key, const uint8_t* m, uint8_t* out) {
    uint8_t state[STREEBOG_BLOCK_SIZE];
    uint8_t round_key[STREEBOG_BLOCK_SIZE];
    streebog_xor(key, m, state);
    memcpy(round_key, key, STREEBOG_BLOCK_SIZE);

    for (int i = 0; i < STREEBOG_ROUNDS; ++i) {
        streebog_spl(state);
        streebog_xor(round_key, &STREEBOG_C[i * STREEBOG_BLOCK_SIZE], round_key);
        streebog_spl(round_key);
        streebog_xor(round_key, state, state);
    }

    memcpy(out, state, STREEBOG_BLOCK_SIZE);
}

static void streebog_gN(const uint8_t* N, const uint8_t* h, const uint8_t* m, uint8_t* out) {
    uint8_t K[STREEBOG_BLOCK_SIZE];
    uint8_t t[STREEBOG_BLOCK_SIZE];
    uint8_t result[STREEBOG_BLOCK_SIZE];

    streebog_xor(h, N, K);
    streebog_spl(K);
    streebog_e(K, m, t);
    streebog_xor(t, h, result);
    streebog_xor(result, m, out);
}

void streebog_hash(const uint8_t* data, size_t len, uint8_t digest[STREEBOG_DIGEST_LENGTH]) {
    uint8_t h[STREEBOG_BLOCK_SIZE] = {0};
    uint8_t N[STREEBOG_BLOCK_SIZE] = {0};
    uint8_t Sigma[STREEBOG_BLOCK_SIZE] = {0};
    uint8_t v512[STREEBOG_BLOCK_SIZE] = {0};
    uint8_t zero_block[STREEBOG_BLOCK_SIZE] = {0};
    uint8_t block[STREEBOG_BLOCK_SIZE];
    uint8_t new_h[STREEBOG_BLOCK_SIZE];

    v512[1] = 0x02U;

    size_t remaining = len;
    while (remaining >= STREEBOG_BLOCK_SIZE) {
        size_t offset = remaining - STREEBOG_BLOCK_SIZE;
        memcpy(block, data + offset, STREEBOG_BLOCK_SIZE);
        streebog_gN(N, h, block, new_h);
        memcpy(h, new_h, STREEBOG_BLOCK_SIZE);
        streebog_add(N, v512, N);
        streebog_add(Sigma, block, Sigma);
        remaining -= STREEBOG_BLOCK_SIZE;
    }

    memset(block, 0, STREEBOG_BLOCK_SIZE);
    if (remaining > 0) {
        memcpy(block + (STREEBOG_BLOCK_SIZE - remaining), data, remaining);
    }
    block[STREEBOG_BLOCK_SIZE - 1] = 0x01U;
    streebog_gN(N, h, block, new_h);
    memcpy(h, new_h, STREEBOG_BLOCK_SIZE);

    uint8_t len_block[STREEBOG_BLOCK_SIZE] = {0};
    uint32_t bit_len = (uint32_t)(remaining * 8U);
    len_block[60] = (uint8_t)(bit_len & 0xFFU);
    len_block[61] = (uint8_t)((bit_len >> 8) & 0xFFU);
    len_block[62] = (uint8_t)((bit_len >> 16) & 0xFFU);
    len_block[63] = (uint8_t)((bit_len >> 24) & 0xFFU);

    streebog_add(N, len_block, N);
    streebog_add(Sigma, block, Sigma);

    streebog_gN(zero_block, h, N, new_h);
    memcpy(h, new_h, STREEBOG_BLOCK_SIZE);
    streebog_gN(zero_block, h, Sigma, new_h);
    memcpy(h, new_h, STREEBOG_BLOCK_SIZE);
    memcpy(digest, h, STREEBOG_DIGEST_LENGTH);
}

File: lab6/streebog.py
from __future__ import annotations

import re
from pathlib import Path
from typing import Sequence, Tuple

BLOCK_SIZE = 64
DIGEST_LENGTH = 64
_ROUNDS = 12


def _load_constants() -> Tuple[bytes, bytes, bytes, Tuple[bytes, ...]]:
    base_dir = Path(__file__).resolve().parent.parent / "lab5" / "server" / "src"
    source = base_dir / "streebog_constants.h"
    try:
        text = source.read_text(encoding="utf-8")
    except FileNotFoundError as exc:
        raise RuntimeError(f"Cannot locate Streebog constants at {source}") from exc

    def parse(name: str) -> bytes:
        pattern = re.compile(
            rf"static const uint8_t {name}\[\] = \{{([^}}]+)\}};",
            re.MULTILINE | re.DOTALL,
        )
        match = pattern.search(text)
        if not match:
            raise RuntimeError(f"Failed to locate constant array {name} in {source}")
        values = [int(value, 16) for value in re.findall(r"0x[0-9a-fA-F]+", match.group(1))]
        return bytes(values)

    pi = parse("STREEBOG_PI")
    tau = parse("STREEBOG_TAU")
    l = parse("STREEBOG_L")
    c = parse("STREEBOG_C")

    if len(pi) != 256:
        raise RuntimeError("Unexpected STREEBOG_PI length")
    if len(tau) != BLOCK_SIZE:
        raise RuntimeError("Unexpected STREEBOG_TAU length")
    if len(l) != BLOCK_SIZE * 8:
        raise RuntimeError("Unexpected STREEBOG_L length")
    if len(c) != _ROUNDS * BLOCK_SIZE:
        raise RuntimeError("Unexpected STREEBOG_C length")

    round_constants = tuple(c[i * BLOCK_SIZE : (i + 1) * BLOCK_SIZE] for i in range(_ROUNDS))
    return pi, tau, l, round_constants


_PI, _TAU, _L, _ROUND_CONSTANTS = _load_constants()


def _xor_bytes(a: Sequence[int], b: Sequence[int]) -> bytearray:
    return bytearray((x ^ y) & 0xFF for x, y in zip(a, b))


def _add_bytes(a: Sequence[int], b: Sequence[int]) -> bytearray:
    carry = 0
    result = bytearray(BLOCK_SIZE)
    for i in range(BLOCK_SIZE):
        total = int(a[i]) + int(b[i]) + carry
        result[i] = total & 0xFF
        carry = total >> 8
    return result


def _s_layer(block: Sequence[int]) -> bytearray:
    output = bytearray(BLOCK_SIZE)
    for i in range(BLOCK_SIZE):
        output[i] = _PI[block[BLOCK_SIZE - 1 - i]]
    return output


def _p_layer(block: Sequence[int]) -> bytearray:
    output = bytearray(BLOCK_SIZE)
    for i in range(BLOCK_SIZE):
        output[i] = block[_TAU[BLOCK_SIZE - 1 - i]]
    return output


def _l_layer(block: Sequence[int]) -> bytearray:
    output = bytearray(BLOCK_SIZE)
    for i in range(7, -1, -1):
        for n in range(8):
            p = 63
            acc = 0
            for j in range(7, -1, -1):
                byte = block[i * 8 + j]
                for k in range(8):
                    if (byte >> k) & 0x01:
                        acc ^= _L[p * 8 + n]
                    p -= 1
            output[i * 8 + n] = acc
    return output


def _spl(block: Sequence[int]) -> bytearray:
    return _l_layer(_p_layer(_s_layer(block)))


def _encrypt(key: Sequence[int], block: Sequence[int]) -> bytearray:
    state = _xor_bytes(key, block)
    round_key = bytearray(key)
    for i in range(_ROUNDS):
        state = _spl(state)
        round_key = _xor_bytes(round_key, _ROUND_CONSTANTS[i])
        round_key = _spl(round_key)
        state = _xor_bytes(round_key, state)
    return state


def _g_transform(N: Sequence[int], h: Sequence[int], m: Sequence[int]) -> bytearray:
    K = _spl(_xor_bytes(h, N))
    t = _encrypt(K, m)
    result = _xor_bytes(t, h)
    return _xor_bytes(result, m)


def hash512(data: bytes) -> bytes:
    h = bytearray(BLOCK_SIZE)
    N = bytearray(BLOCK_SIZE)
    Sigma = bytearray(BLOCK_SIZE)
    v512 = bytearray(BLOCK_SIZE)
    v512[1] = 0x02
    zero_block = bytes(BLOCK_SIZE)

    remaining = len(data)
    while remaining >= BLOCK_SIZE:
        offset = remaining - BLOCK_SIZE
        block = bytearray(data[offset : offset + BLOCK_SIZE])
        h = _g_transform(N, h, block)
        N = _add_bytes(N, v512)
        Sigma = _add_bytes(Sigma, block)
        remaining -= BLOCK_SIZE

    block = bytearray(BLOCK_SIZE)
    if remaining > 0:
        block[BLOCK_SIZE - remaining :] = data[:remaining]
    block[-1] = 0x01
    h = _g_transform(N, h, block)

    len_block = bytearray(BLOCK_SIZE)
    bit_len = (remaining * 8) & 0xFFFFFFFF
    len_block[60] = bit_len & 0xFF
    len_block[61] = (bit_len >> 8) & 0xFF
    len_block[62] = (bit_len >> 16) & 0xFF
    len_block[63] = (bit_len >> 24) & 0xFF

    N = _add_bytes(N, len_block)
    Sigma = _add_bytes(Sigma, block)

    h = _g_transform(zero_block, h, N)
    h = _g_transform(zero_block, h, Sigma)
    return bytes(h)


__all__ = ["hash512", "BLOCK_SIZE", "DIGEST_LENGTH"]
